src/main.rs

```diff
@@ -1,183 +1,153 @@
 #![feature(panic_info_message)]
 
-#[macro_use] extern crate failure;
 #[macro_use] extern crate serde_derive;
 
-mod format;
 mod lookup;
 mod models;
 mod utils;
 
-use self::models::*;
+use models::ExtendSelectionParams;
 
-use failure::Error;
-use lsp_types::*;
-use rnix::{parser::*, types::*};
+use log::{error, trace, warn};
+use lsp_server::{Connection, ErrorCode, Message, Notification, Request, RequestId, Response};
+use lsp_types::{
+    *,
+    request::{*, Request as _}
+};
+use rnix::{parser::*, types::*, SyntaxNode};
 use std::{
     collections::HashMap,
-    fmt,
-    fs::File,
-    io::{self, prelude::*},
     panic,
-    rc::Rc
+    rc::Rc,
 };
 
-fn main() -> Result<(), Error> {
-    let mut log = File::create("/tmp/nix-lsp.log")?;
-    let log_clone = log.try_clone()?;
+pub type Error = Box<dyn std::error::Error>;
 
-    let stdout = io::stdout();
-    let mut app = App {
-        files: HashMap::new(),
-        log: &mut log,
-        stdout: stdout.lock()
-    };
+pub const DUMMY_ERROR: &str = "A fatal error has occured and nix-lsp has shut down.";
+
+fn main() -> Result<(), &'static str> {
```

Comment:
Note that `main() -> Result<(), &str>` will use `fmt::Debug` to print `&str`, so, for best results, it's still better to print the error manually.

Reply:
Right, I'm mainly wishing to return the right exit code. I could do it using `process::exit` but... destructors :(

Reply:
I like how ripgrep does this, very tidy: https://github.com/BurntSushi/ripgrep/blob/master/src/main.rs#L48

Reply:
That's somewhat what I wanted to avoid, but fine :)